### PR TITLE
misc: refine vuart connection

### PR DESCRIPTION
**Pull request by weichenli-intel**
We have redesign the vuart and the UI for user, so the config tool
should change the schema and xform for the new xml, then change the
static_allocators to alloc irq and io_port for new connection.

This patch add a new vuart connection type and change the xforms and xmls
to adapter the new design.

This patch dependent on the change of one source patch.

Tracked-On: #6690
Reviewed-by: Junjie Mao <junjie.mao@intel.com>
Signed-off-by: Chenli Wei <chenli.wei@linux.intel.com>

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

**Comment by szhen11** [comment]
@weichenli-intel  build check for some boards failed, please update the config xmls for these scenario xmls and add Depend-On tag.

**Comment by weichenli-intel** [comment]
> @weichenli-intel build check for some boards failed, please update the config xmls for these scenario xmls and add Depend-On tag.

Yes, "This patch dependent on the change of one source patch."
I will update these xml by another channel.

File: misc/config_tools/data/nuc11tnbi5/shared.xml [hunk 1]
```diff
@@ -52,6 +52,32 @@
...
+          <vm_name>POST_STD_VM1</vm_name>
+          <io_port>0x3F</io_port>
+        </endpoint>
+      </vuart_connection>
+      <vuart_connection>
+        <name>Connection_2</name>
+        <type>pci</type>
+        <endpoint>
+          <vm_name>ACRN_Service_VM</vm_name>
+          <vbdf>00:10.1</vbdf>
```
Comment (junjiemao1) [inline misc/config_tools/data/nuc11tnbi5/shared.xml:73]: Should this be `00:10.0`? Without function 0 Linux will not recognize other functions of the same device.

**Review by NanlinXie — APPROVED** [review]

**Comment by szhen11** [comment]
Build doc failed:
making HTML content
doc build tool versions found on your system...

  breathe                   version: 4.31.0
  sphinx                    version: 3.5.4
  docutils                  version: 0.16
  sphinx-rtd-theme          version: 1.0.0
  sphinx-tabs               version: 1.3.0
  doxygen                   version: 1.8.17


New errors/warnings found, please fix them:
==============================================

/home/acrn/workspace/acrn-hypervisor-build-doc_155/acrn-hypervisor/build/doc/rst/release_notes/release_notes_2.7.rst:147: WARNING: unknown option: vm.legacy_vuart.base
/home/acrn/workspace/acrn-hypervisor-build-doc_155/acrn-hypervisor/build/doc/rst/release_notes/release_notes_2.7.rst:147: WARNING: unknown option: vm.legacy_vuart.irq
/home/acrn/workspace/acrn-hypervisor-build-doc_155/acrn-hypervisor/build/doc/rst/tutorials/nvmx_virtualization.rst:182: WARNING: unknown option: vm.legacy_vuart.base

make[1]: *** [Makefile:64: html] Error 1

**Comment by dbkinder** [comment]
PR #7177 fixes the doc build error caused by removing options that are referenced in the documentation.  If you rebase you should pick up that change and remove the doc build error.

**Comment by szhen11** [comment]
@weichenli-intel  premerge test failed when launching Service VM by "vm_console 0":
[2022-03-03 14:57:43.148] [[0;36m INFO [0m] Enter service vm console: vm_console 0
[ ERROR ] write: vm_console 0
 and expect: login: failed with timeout: 90
vm_console 0

vuart console is not active 

ACRN:\>

ACRN:\>

**Comment by dbkinder** [comment]
This change does require an update to ``tutorials/nvmx_virtualization.rst:`` that references settings to ``vm.legacy_vuart.base``, an option that no longer exists.

**Review by NanlinXie — APPROVED** [review]

**Comment by weichenli-intel** [comment]
> 

resolved

File: misc/config_tools/schema/config.xsd [hunk 2]
```diff
@@ -298,6 +298,12 @@ If this value is empty, then the default value will be calculated from informati
     <xs:element name="MEMORY" type="MemoryOptionsType" />
     <xs:element name="CAPACITIES" type="CapacitiesOptionsType" />
     <xs:element name="MISC_CFG" type="MiscCfgOptionsType" />
+    <xs:element name="vuart_connections" type="VuartConnectionsType" minOccurs="0" maxOccurs="8">
+      <xs:annotation>
+        <xs:documentation>Specify the vUART connection setting.
+Refer to :ref:`vuart_config` for detailed vUART settings.</xs:documentation>
```
Comment (gvancuts) [inline misc/config_tools/schema/config.xsd:304]: @dbkinder does this `:ref:` mechanism work from within an `xsd` file?

File: misc/config_tools/service_vm_config/serial_config.py [hunk 3]
```diff
@@ -37,20 +37,6 @@ def main(args):
                 user_vm_id = legacy_vuart.find('target_vm_id').text
                 legacy_vuartid = int(legacy_vuart.attrib["id"])
                 vuart_target_vmid[legacy_vuartid] = user_vm_id
-
-    vm_list = allocation_etree.xpath("//vm[load_order = 'SERVICE_VM']")
-    for vm in vm_list:
-        vuart_list = find_non_standard_uart(vm)
```
Comment (gvancuts) [inline misc/config_tools/service_vm_config/serial_config.py:43]: Does this patch remove the generation of the `serial.conf` file that we use to expose non-standard serial port in the Service VM?

Comment (weichenli-intel) [reply misc/config_tools/service_vm_config/serial_config.py:43]: We have not removed the support for non-standard serial port. 
Users could set non-standard UART by the new UI, IRQ is generated automatically, and io port use a default value.

Comment (gvancuts) [reply misc/config_tools/service_vm_config/serial_config.py:43]: But there is no need for the `serial.conf` file ever? The Service VM will always see whatever serial port the user sets via the UI, is that correct?

**Comment by NanlinXie** [comment]
Doc task tracked in github issue:
https://github.com/projectacrn/acrn-hypervisor/issues/6690

File: misc/config_tools/service_vm_config/serial_config.py (hunk 3 above)
Comment (gvancuts) [reply misc/config_tools/service_vm_config/serial_config.py:43]: > 

Can you please answer this?

Comment (weichenli-intel) [reply misc/config_tools/service_vm_config/serial_config.py:43]: I have check again, now our vuart driver can't create non-stand vuart.
So the SOS still need the serial.conf.

But the logic of create serial.conf can't adapter the new vuart connection, so these logic should update.
There will be another patch to fix this.

It's a great honor to invite you to review these patch.

**Pull request merged**